Add secured private asset transfer example to docs (#1541)
hyperledger/fabric: docs/source/private-data/private-data.md

  for example to specify that only an owner's organization peer, custodian's organization
  peer, or other third party must endorse such transactions.

## Example scenario: Asset transfer using private data collections

The private asset transfer smart contract demonstrates how an asset can be represented and traded between organizations in a Hyperledger Fabric blockchain channel, while keeping details of the asset and transaction private.
Each on-chain asset is a non-fungible token (NFT) that represents a specific asset having certain immutable metadata properties (such as size and color) with a unique owner. When the owner wants to sell the asset, both parties need to agree to the same price before the asset is transferred. The private asset transfer smart contract enforces that only the owner of the asset can transfer the asset. In the course of this tutorial, you will learn how Fabric features such as state based endorsement, private data, and access control come together to provide secured transactions that are both private and verifiable.

### Scenario requirements

> The private asset transfer scenario is bound by the following requirements:

Reviewer: if this whole page relates to specifically `secured-agreement` sample, we should provide the link somewhere on this page : https://github.com/hyperledger/fabric-samples/tree/master/asset-transfer-secured-agreement/chaincode-go

my main motivation being that this page  titled `private data collections` is easily confused with the `asset-transfer-private-data`  sample, being worked on here https://github.com/hyperledger/fabric-samples/pull/205

Author: Good point, I added back the original scenario section and added a new section for this tutorial (including a link).

Author: As suggested by others, converting to a separate tutorial.

- A asset may be issued by the first owner's organization (in the real world issuance may be restricted to some authority that certifies a asset's properties).
- Ownership is managed at the organization level (the Fabric permissioning scheme would equally support ownership at an individual identity level within an organization).
- The asset identifier and owner is stored as public channel data for all channel members to see.
- The asset metadata properties however are private information known only to the asset owner (and prior owners).

> - An interested buyer will want to verify a asset's private properties.

Reviewer: nit: `an asset` in couple of places

Author: 🛠️

And instead of transactors directly hosting peers, they could transact through
custodian organizations who are running peers.

## Example scenario: Secured Private Asset transfer

> The private asset transfer smart contract [fabric sample](https://github.com/hyperledger/fabric-samples/tree/master/asset-transfer-secured-agreement/chaincode-go) 

Reviewer 2: We have 2 example scenarios in here (which is fine) but one is at an h3 ### and the other is at at h2 ##.

Would it make more sense to have:
h2 ## Example scenarios
.. introduce and explain the difference between them and why we have included 2 here ....

h3 ### Example scenario 1: Asset transfer using private data collections
h3 ### Example scenario 2: Secured Private Asset transfer

Author: As suggested by others, converting to a separate tutorial.

- A hash of each price agreement is stored on the ledger. The two hashes will match only if the two organizations have agreed to the same price. This allows the organizations to verify that they have come to agreement on the transfer details before the transfer takes place. A random trade id is added to the price agreement, which serves as a *salt* to ensure that other channel members can not use the hash on the ledger to guess the price.

**Step 3: Transferring the asset**

After the two organizations have agreed to the same price, the asset owner can use the transfer function to transfer the asset to the buyer:

- Smart contract access control ensures that the transfer must be initiated by a member of the organization that owns the asset.

> - The transfer function verifies that the asset details passed to the transfer function matches the on chain hash, to ensure that the asset owner is *selling* the same asset that they own.

Reviewer: ```suggestion
- The transfer function verifies that the asset's private immutable properties passed to the transfer function matches the on chain hash of the asset data in private collection, to ensure that the asset owner is *selling* the same asset that they own.
```

Author: 🛠️

- The state based endorsement policy of the public asset record is updated so that only a peer of the new owner of the asset can update or sell their new asset.

### Running the private asset transfer smart contract

You can use the Fabric test network to run the private asset transfer smart contract. The test network contains two peer organizations, Org1 and Org1, that operate one peer each. In this tutorial, we will deploy the smart contract to a channel of the test network joined by both organizations. We will first create a asset that is owned by Org1. After the two organizations agree on a asset price, we will transfer the asset from Org1 to Org2.

### Deploy the test network

We need to deploy an instance of the Fabric test network to run the smart contract. Open a command terminal and navigate to test network directory in your local clone of [fabric-samples](https://github.com/hyperledger/fabric-samples). We will operate from the `test-network` directory for the remainder of the tutorial.
```
cd fabric-samples/test-network
```

You can then deploy the network with the following command:
```

> ./network.sh up createChannel

Reviewer: since new developers use the same network for our sample application testing as well, I think our sample applications need CA to be part of  the network (regiserUser with ca etc) and use it for cryptogen. If this is not true, I stand corrected

Making channel name explicit as well.
```suggestion
./network.sh up createChannel -ca -c mychannel
```

Author: Added the channel name; I did not include `-ca` as the sample is written as a standalone chaincode and does not perform any new identity registration.

#### Set the environment variables to operate as Org2

Now that we have one terminal that we can operate as Org1, open a new terminal for Org2. Make sure that this terminal is also operating from the `test-network` directory. Set the following environment variables to operate as the Org2 admin:
```
export PATH=${PWD}/../bin:${PWD}:$PATH
export FABRIC_CFG_PATH=$PWD/../config/
export CORE_PEER_TLS_ENABLED=true
export CORE_PEER_LOCALMSPID="Org2MSP"
export CORE_PEER_MSPCONFIGPATH=${PWD}/organizations/peerOrganizations/org2.example.com/users/Admin@org2.example.com/msp
export CORE_PEER_TLS_ROOTCERT_FILE=${PWD}/organizations/peerOrganizations/org2.example.com/peers/peer0.org2.example.com/tls/ca.crt
export CORE_PEER_ADDRESS=localhost:9051
```

You will need switch between the two terminals as you go through the tutorial.

### Deploy the chaincode

Now that we can operate as both organizations, we need install the private asset transfer smart contract on the peers of Org1 and Org2, and deploy the chaincode to the channel approving and committing the chaincode definition.

#### Install and approve the chaincode as Org1

> Open the Org1 terminal. Run the following command to package the private asset transfer chaincode:

Reviewer: for golang, before package, I think we should add that:

package your chaincode dependencies using:
```
go mod tidy
go mod vendor
```

Author: I agree tidying is a good practice but I don't believe it's a useful addition to the sample. On `go mod vendor`, why? This sample does not have a vendor folder.

Reviewer: I take mod vendor back. For golang , deployCC,sh does call that to package the go dependency in vendor folder. So no need.

Reviewer 3: In 2.x it doesn't matter. If a `go.mod` is present the Peer will explicitly enter module mode at compilation for you. If a `vendor` directory is present the peer will explicitly disable module mode.

Vendoring is useful in places where the internet is not necessarily particularly stable, for example China/India. It's not uncommon to get timeouts when installing chaincode when the dependencies aren't included as part of the vendor folder and Go has to fetch them. This is a frequent complaint from community members outside North America.

```
peer lifecycle chaincode package assets_transfer.tar.gz --path ../asset-transfer-secured-agreement/chaincode-go --lang golang --label assets_transfer_1
```

The command creates a chaincode package named `assets_transfer.tar.gz`. We can now install this package on the Org1 peer:
```
peer lifecycle chaincode install assets_transfer.tar.gz
```

You will need the chaincode package ID in order to approve the chaincode definition. You can find the package ID by querying your peer:
```
peer lifecycle chaincode queryinstalled
```
Save the package ID as an environment variable. The package ID will not be the same for all users, so need to use the result that was returned by the previous command:
```
export PACKAGE_ID=assets_transfer_1:2a585633baa0a6ba0019965ac40d6f188194c50df1015010b080ef6ba426d266
```
You can now approve the chaincode as Org1:
```

> peer lifecycle chaincode approveformyorg -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --channelID mychannel --name assets_transfer --version 1 --package-id $PACKAGE_ID --sequence 1 --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem --signature-policy "OR('Org1MSP.peer','Org2MSP.peer')"

Reviewer: A more explicit chaincode name in all the commands will ensure that all our asset_transfer_* samples can be installed at same time
```suggestion
peer lifecycle chaincode approveformyorg -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --channelID mychannel --name assets_transfer_secure --version 1 --package-id $PACKAGE_ID --sequence 1 --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem --signature-policy "OR('Org1MSP.peer','Org2MSP.peer')"
```

btw, once we have new short names in test-network, per https://github.com/hyperledger/fabric-samples/pull/235 , we should add a `QuickInstall` instruction using the chaincode short name called `secure`. I think having the detailed steps like this is still useful for devops to do custom steps.

Author: 🛠️ wording; Good idea, we can add `QuickInstall` in a follow-up pr.

```
export asset_PRICE=$(echo -n "{\"asset_id\":\"asset1\",\"trade_id\":\"109f4b3c50d7b0df729d299bc6f8e9ef9066971f\",\"price\":110}" | base64)
peer chaincode invoke -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem -C mychannel -n assets_transfer -c '{"function":"AgreeToSell","Args":["asset1"]}' --transient "{\"asset_price\":\"$asset_PRICE\"}"
```

We can query the Org1 private data collection to read the agreed to selling price:
```
peer chaincode query -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem -C mychannel -n assets_transfer -c '{"function":"GetAssetSalesPrice","Args":["asset1"]}'
```

### Agree to buy as Org2

Operate from the Org2 terminal. Run the following command to verify the asset properties before agreeing to buy. The asset properties and salt would be passed out of band, through email or other communication, between the buyer and seller.
```

> export asset_PROPERTIES=$(echo -n "{\"object_type\":\"asset_properties\",\"asset_id\":\"asset1\",\"color\":\"blue\",\"size\":35,\"salt\":\"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3\"}" | base64)

Reviewer 4: Let's make variables like "asset_PROPERTIES" all caps throughout, I think this was a copy/paste mistake in the original tutorial.

Also, this one should be a "peer chaincode query" rather than "peer chaincode invoke".

Author: 🛠️
